complete todo app

arfedulov opened the pull request.

arfedulov committed:
Task component

arfedulov committed:
TaskFilter component

arfedulov committed:
fix(Task): change name

arfedulov committed:
fix(TaskFilter): add name

arfedulov committed:
add Toolbar component

arfedulov committed:
add TaskList component

arfedulov committed:
store: implement  mutations

arfedulov committed:
test: store actions

arfedulov committed:
localStorage api

arfedulov committed:
implement TaskList component

arfedulov committed:
implement filter function on TaskList

arfedulov committed:
TaskList: connect with task-filter

arfedulov committed:
fix: keep $api on store instance

arfedulov committed:
TaskList: add tollbar

arfedulov committed:
TaskList: connect createTask to Toolbar

arfedulov committed:
fix: createTask api call return created task

arfedulov committed:
Task: is editable only on focus

arfedulov committed:
fix(Task): emit full task data on update

arfedulov committed:
fix: filter tasks by done status

arfedulov committed:
validate date

arfedulov committed:
fix(localStorageApi): parse dates

arfedulov committed:
add bootstrap styles

arfedulov committed:
fix:tests dont fail because of using bootstrap-vue

In unit tests for components that use bootstrap-vue
    1) change shallowMount to mount;
    2) use localVue with BootstrapVue plugin;
    3) use `input` tag selector for for finding input element

arfedulov committed:
add form validation

artsem-teal commented on src/App.vue:
```diff
@@ -1,32 +1,19 @@
-}
+export default {
+  components: {
+    'task-list': TaskList,
```
Можно не писать названия в kamel-case, когда регистрируешь компонент. В темплейте будет воспринимать как CamelCase, так и kebab-case.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+<template>
+  <b-card tag="article" :bg-variant="done ? 'success' : 'default'">
```
done ? 'success' : 'default' лучше вынести в computed.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+    <b-form-checkbox class="task-status-control" v-model="statusModel" />
+    <b-card-title>
+      <b-form-input
+        :state="!errors.title.length"
```
Это тоже можно в computed.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+      </b-form-invalid-feedback>
+    </b-card-title>
+    <b-card-text>
+      <b-form-textarea v-if="editable" class="task-content-control" v-model="contentModel" />
```
По названию contentModel. В названиях таких переменных лучше не привязываться к каким-то специфичным в темплейте вещам. Не очень понятно что и для чего.

arfedulov replied:
Ты имеешь в виду, что название `contentModel` содержит в себе `model`? т.е. здесь в название переменной закодировано ее отношение к директиве `v-model`. Лучше поменять на более абстрактное название `content`?

artsem-teal replied:
Да. У тебя же в переменной хранятся какие-то данные. Модель никак не относится к этим данным, кроме как тем, что ты кидаешь эти данные сюда.

Если бы ты использовать через value и обработчик эвента - не было бы никаких проблем. Потому что ты бы в value ставил значение из пропса, а событие эмитил бы наверх.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+    </b-card-text>
+    <footer>
+      <b-button-group>
+        <b-button variant="danger" class="task-remove-control" @click="remove">delete</b-button>
```
Названия методам лучше давать более понятные. Тут, например, можно написать removeTask. В данном варианте не очень понятно, что оно удалит.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+    <footer>
+      <b-button-group>
+        <b-button variant="danger" class="task-remove-control" @click="remove">delete</b-button>
+        <b-button variant="warning" v-if="!editable" @click="editable=true">edit</b-button>
```
Вынеси в отдельный метод лучше.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+      <b-button-group>
+        <b-button variant="danger" class="task-remove-control" @click="remove">delete</b-button>
+        <b-button variant="warning" v-if="!editable" @click="editable=true">edit</b-button>
+        <b-button variant="primary" v-else @click="onSubmit">save</b-button>
```
Я бы лучше написал saveChanges или что-то такое.

arfedulov replied:
Здесь еще сложность в том, что при нажатии на кнопку на самом деле сохранения не происходит. Сохранение происходит автоматом по мере печатания. Может лучше тогда переименовать в "stop editing"?

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+  data() {
+    return {
+      editable: false,
+      errors: {
```
Если у тебя больше не будет в объекте свойств, то лучше просто сделать свойство titleErrors.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+    };
+  },
+  computed: {
+    titleModel: {
```
В принципе, такой способ тоже рабочий, но я бы предпочел вместо v-model использовать :value и соответствующий эвент. Потому что такой вариант сложнее читается.

p.s. Просто на будущее, можешь не исправлять. :)

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+      this.$emit(EVENTS.TASK_DELETE, this.id);
+    },
+    setStatus(done) {
+      if (done) {
```
Давай тут тернарный оператор используем.

arfedulov replied:
Если использовать тернарный оператор, eslint ругается на экспрешн без присвоения.

artsem-teal replied:
А ты сохраняй название эвента, а ниже просто прокидывай его в emit.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+      this.errors = {
+        title: [],
+      };
+      if (this.title !== '' && this.title.length <= 25) {
```
Я бы немного по-другому написал.

Вот этот if убрать и по умолчанию возвращать true из функци.
А уже в случаях ошибки добавлять ошибку и возвращать false.

artsem-teal commented on src/components/TaskFilter.vue:
```diff
@@ -0,0 +1,89 @@
+  if (!date) {
+    return '';
+  }
+  return `${date.getFullYear()}-${date.getMonth().toString().padStart(2, '0')}-${date.getDate().toString().padStart(2, '0')}`;
```
Давай разнесем вычисления по переменным, а тут уже просто склеим их.

artsem-teal commented on src/components/TaskFilter.vue:
```diff
@@ -0,0 +1,89 @@
+export default {
+  name: 'task-filter',
+  props: {
+    done: Boolean,
```
Рекомендуется писать либо default либо required. Чтобы мы могли гарантировать, что будет хоть какое-то значение.

artsem-teal commented on src/components/TaskFilter.vue:
```diff
@@ -0,0 +1,89 @@
+  },
+  methods: {
+    toggleDone(done) {
+      if (!done) {
```
Тернарный оператор

arfedulov replied:
Здесь, как и выше, eslint ругается на экспрешн без присвоения.

artsem-teal replied:
Ответил выше.

artsem-teal commented on src/components/Toolbar.vue:
```diff
@@ -0,0 +1,26 @@
+
+<script>
+export const EVENTS = {
+  CREATE_TASK: 'toolbar:create-task',
```
Можно просто в одиночную константу вынести.

arfedulov replied:
А я просто во всем проекте такую схему использую. Поэтому, думал, что стоит оставаться последовательным и хранить названия эвентов в объекте `EVENTS`.

artsem-teal commented on src/localStorageApi.js:
```diff
@@ -0,0 +1,68 @@
+        .map(parseTask);
+      const task = createTask();
+      tasks.push(task);
+      localStorage.setItem('tasks', JSON.stringify(tasks));
```
Item локал сторэджа можно вынести в константу, чтобы использовать ее везде.

artsem-teal commented on src/store/index.js:
```diff
@@ -1,15 +1,94 @@
+    if (taskIndex === -1) {
+      throw new Error(`Task with id ${task.id} does not exist`);
+    }
+    Vue.set(state.tasks, taskIndex, { ...state.tasks[taskIndex], ...task });
```
А почему тут Vue.set использовал?

arfedulov replied:
Насколько я понял, мы не можем мутировать `state.tasks` массив (например методом `Array.prototype.splice`) т.к. в этом случае изменения пройдут незаметно для реактивной системы. Если я не заблуждаюсь в этом, то получается, что мутировать `state.tasks` можно либо так, как я сделал, либо просто присваивая измененный массив
```js
state.tasks = state.tasks.reduce((acc, oldTask, i) => {
    if(i === taskIndex) {
        acc.push({ ...oldTask, ...task });
    } else {
        acc.push(oldTask);
    }
    return acc;
});
```
Но вариант с `Vue.set` мне как-то больше нравится.

artsem-teal commented on src/localStorageApi.js:
```diff
@@ -0,0 +1,68 @@
+let isFresh = false;
+
+export const api = {
+  getTasks: async () => {
```
А почему async используешь в методах?
Ты же с промисами никак тут не работаешь.

arfedulov replied:
По задумке интерфейс api сервиса возвращает промис. Я использовал `async` как альтернативу возвращения из функции `Promise.resolve(value)`. Насколько я помню, `async` функция в любом случае вернет промис, даже если я из нее возвращаю примитивное значение?

artsem-teal replied:
Слишком перемудрил. Тут же промис совсем не нужен. У тебя асинхронный код.
А делать в данном случае асинхронный - я не понимаю зачем.

artsem-teal commented on src/store/index.js:
```diff
@@ -1,15 +1,94 @@
+    try {
+      if (!await this.$api.isFresh()) {
+        const tasks = await this.$api.getTasks();
+        context.commit('loadTasks', tasks);
```
Можно commit деструкторизировать в аргументах функции.

artsem-teal commented on src/store/index.js:
```diff
@@ -1,15 +1,94 @@
   },
 });
+
+store.$api = api;
```
А почему так? Почему просто не использовать экспортированный объект?

arfedulov replied:
Это мой кривой вариант инъекции апи сервиса в store. В данном случае я использую реализацию `localStorageApi` как дефолтную. При этом у меня есть возможность в тестах очень просто сделать инъекцию мок апи. Можно, наверное, было положиться на `jest.mock()`. Но вариант с инъекцией мне больше нравится. Хотя, что касается инъекции зависимостей во vue, я так и не разобрался с этим вопросом. Как я понял, это не особо популярный паттерн во vue.

В итоге. В данном конкретном случае я использовал такой подход, чтобы в тестах вставлять мок апи.

artsem-teal replied:
Да, так никто не делает.

В тестах, на сколько я помню, можно мокать импорты. Попробуй про это почитать.

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,125 @@
+
+<script>
+export const EVENTS = {
+  TASK_CHANGE: 'task:change',
```
Как вариант, я бы писал названия эвентов типа change-task. Но это уже дело вкуса.

artsem-teal commented on src/views/TaskList.vue:
```diff
@@ -0,0 +1,124 @@
+      <task
+        v-for="task of filteredTasks" :key="task.id"
+        class="task"
+        :id="task.id"
```
Почитай про v-bind и как тут его можно использовать.

artsem-teal commented on src/views/TaskList.vue:
```diff
@@ -0,0 +1,124 @@
+import Toolbar from '@/components/Toolbar.vue';
+
+// eslint-disable-next-line no-restricted-globals
+const isValidDate = date => date && !isNaN(date.getTime());
```
Я так полагаю он ругается на isNaN? Давай послушаем eslint и напишем Number.isNaN. :)

artsem-teal commented on src/views/TaskList.vue:
```diff
@@ -0,0 +1,124 @@
+// eslint-disable-next-line no-restricted-globals
+const isValidDate = date => date && !isNaN(date.getTime());
+
+export const filterTasks = (tasks, filters) => {
```
Ох. Давай без лишних экспортов из компонентов. Лучше вынести в какой-нибудь utils.

И как-то очень сложно такой метод вообще прочитать. Порефактори его как-нибудь. :)

artsem-teal commented on src/views/TaskList.vue:
```diff
@@ -0,0 +1,124 @@
+  name: 'task-list',
+  components: {
+    task: Task,
+    'task-filter': TaskFilter,
```
Не обязательно тут писать в kebab. Уже сверху упоминал.

artsem-teal commented on src/views/TaskList.vue:
```diff
@@ -0,0 +1,124 @@
+  },
+  computed: {
+    tasks() {
+      return this.$store.state.tasks;
```
Используй mapState и т.п. из библиотеки vuex.

artsem-teal commented on src/views/TaskList.vue:
```diff
@@ -0,0 +1,124 @@
+    },
+  },
+  created() {
+    this.$store.dispatch('syncTasks');
```
Напрямую никогда так не делай. Используй мапперы.

artsem-teal commented on tests/unit/Task.spec.js:
```diff
@@ -0,0 +1,140 @@
+localVue.use(BootstrapVue);
+
+describe('Task', () => {
+  it('emits `task:change` event on `title` change with updated `title` and previous `content` as a payload', () => {
```
Слишком длинное название. Я думаю "emits `task:change` event on `title` change" достаточно. Не нужно слишком много описывать в названии.

arfedulov replied:
Здесь я хотел полностью описать требуемое поведение. Выглядит действительно очень длинно и неудобно. Может быть есть какой-то более чистый способ писать подробные спецификации? Просто в данном случае, если я сокращу описание, то следующий кейс
будет иметь точно такое же описание, как и этот кейс. Получается, что не понятно, что мы тестируем.

artsem-teal commented on tests/unit/Task.spec.js:
```diff
@@ -0,0 +1,140 @@
+    const EXPECT = {
+      title: TITLE,
+      content: CONTENT,
+      updatedAt: new Date(),
```
Могут быть проблемы с new Date, так как ты не знаешь, когда он выполнится относительно 17-й строки.

artsem-teal commented on tests/unit/Task.spec.js:
```diff
@@ -0,0 +1,140 @@
+    const TITLE = 'qwert';
+
+    wrapper.vm.editable = true;
+    wrapper.find('.task-title-control').setValue(TITLE);
```
Может лучше ставить значение в переменную, а не использовать темплейт для этого?

arfedulov replied:
Я не понял вопроса.

artsem-teal replied:
Тут, по сути, не очень хороший тест. Ты проверяешь работу v-model - а именно ставишь значение в дом элемент, чтобы потом сработал set компьютеда, чтобы тот вызвал функцию.

Если эта функция отдает наружу что-то - нужно тестировать эту функцию.

Вообще, нужно тестировать только внешнее поведение компонентов, а ты тут слишком много тестируешь сам фреймворк, получается.

arfedulov committed:
fix(Task.spec): do not rely on exact date in tests

arfedulov committed:
use short way for declaring  in component options

arfedulov committed:
move to computed property Card's background variant

arfedulov committed:
move to computed props title input validation status

arfedulov committed:
rename `remove` method to `removeTask`

arfedulov committed:
refactor inline expression in template into method

arfedulov committed:
rename `save` button to `stop editing`

arfedulov committed:
keep title errors in `titleErrors`  array

arfedulov committed:
refactor form validation method

Массив this.titleErrors присваивается в любом
случае, т.к. необходимо очистить его от ошибок
предыдущей валидации. Т.к. теоретически несколько
if веток могут исполниться и добавить ошибку в массив,
я не возвращаю false сразу из тела if.

arfedulov committed:
refactor formatDate method

arfedulov committed:
use default props

arfedulov committed:
use ITEM_KEY for localStorage

arfedulov committed:
destructure context argument in actions

arfedulov committed:
use short way for binding properties to Task

arfedulov committed:
call Number.isNaN explicitly

arfedulov committed:
move filterTask utility function into separate file

arfedulov committed:
use mapState

arfedulov committed:
use mapActions

arfedulov committed:
reduce test case title length

artsem-teal commented on src/components/Task.vue:
```diff
@@ -0,0 +1,127 @@
+export default {
+  name: 'task',
+  props: {
+    id: { type: Number, required: true },
```
Чтобы избежать проблем с именами - можешь в этот компонент передавать целый объект task, а в нужных местах уже брать нужные свойства.

arfedulov replied:
т.е. я делаю так?

```js
props: {
    task: { type: Object, required: true },
},
```
Но в таком случае я лишаюсь валидации.

arfedulov committed:
dont encode v-model into propery names

Переименовал computed props, но продолжил
использовать v-model директиву, т.к. я исполььзую
разные инпуты (checkbox, text etc.) и v-model
директива позволяет не думать о различиях
в пропсах для разных инпутов.

arfedulov committed:
use ternary operator instead of if statement

arfedulov committed:
dont use async functions in localStorageApi

arfedulov committed:
do not use DI for api in store

arfedulov committed:
refactor test: test only component, not framework
